substratevm/DEBUGINFO.md

```diff
@@ -0,0 +1,207 @@
+sources/src corresponds to the top level package for the classes whose
+sources are being included.
+
+You can also add extra directories to the search path. Note that gdb
+does not understand zip fomrat file systems so any extra entries you
+add must identify a directory tree containing the relevant
+sources. Once again. top leel entries in the directory added to the
+search path must correspond to the top level package for the classes
+whose sources are being included.
+
+Configuring source paths in VS
+------------------------------
+
+TO BE ADDED
+
+Checking debug info on Linux
+----------------------------
+
+n.b. this is only of interest to those who want to understand how the
+debug info implemetation works or want to trouble shoot problems
+encountered during debugging that might relate to the debug info
+encoding.
+
+The objdump command can be used to display the dbeug info embedded
+into a native image. The following commands (which all assume the
+target binary is called hello) can be used to display all currentyl
+generated content:
+
+    $ objdump --dwarf=info hello > info
+    $ objdump --dwarf=abbrev hello > abbrev
+    $ objdump --dwarf=ranges hello > ranges
+    $ objdump --dwarf=decodedline hello > decodedline
+    $ objdump --dwarf=rawline hello > rawline
+    $ objdump --dwarf=str hello > str
+    $ objdump --dwarf=frames hello > frames
+
+The *info* section includes details of all compiled Java methods.
+
+The *abbrev* sectio defines the layout of records in the info section
+that describe Java files (compilation units) and methods.
+
+The *ranges* section details the start and end addresses of method
+code segments
+
+The *decodedline* section maps subsegments of method code range
+segments to files and line numbers. This mapping includes entries
+for files and line numbers for inlined methods.
+
+The *rawline* segment provides deatails of how the line table is
+generated using DWARF state machine instuctions that encode file,
+line and address transitions.
+
+The *str* section provides a lookup table for strings referenced
+from records in the info section
+
+The *frames* section lists transition points in compiled methods
+where a (fixed size) stack frame is pushed or popped, allowing
+the debugger to identify each frame's current and previous stack
+pointers and it's return address.
+
+Note that some of the content embedded in the debug records is
+generated by the C compiler and belongs to code that is either in
+libraries or the C lib bootstrap code that is bundled in with the
+Java method code.
+
+Currently supported targets
+---------------------------
+
+The prototype is currently implemented only for gdb on Linux.
+
+  - Linux/x86_64 suppoort has been tested and should work
+    correctly.
+
+  - Linux/AArch64 support is present but has not yet been fully
+    verified (break points should work ok but stack backtraces
+    may be incorrect).
+
+Windows support is still under development.
```

Comment:
> Windows support is still under development.

It would be useful to also see the development of the Windows support happening in this same repo.

Reply:
Yes, I was thinking the same myself just yestrday.

I believe my colleague Simon Tooke has some basic things working but is is still working on getting a stable version that he is happy to publish. He was struggling with getting VS to use the right file paths but the sources cache may have simplified that. I'll try to push him to provide a version I can pull into this PR branch that will build and maybe implement some basic debugging (I am sure that can be made to happen very soon ;-).

Of course, Simon's code is always lagging slightly behind my version because he keeps having to play catch up with the changes I have been adding (sources cache, Streams in native image <-> object API, other tweaks to API). So, this may take a day or two.

Reply:
> I'll try to push him to provide a version I can pull into this PR branch that will build and maybe implement some basic debugging

That would be great!

Reply:
Simon suggested raising his own PR from a branch in his github repo derived from my branch. That will avoid his changes breaking my branch and vice versa. He can pull my changes in and then publish them once he has assimilated them. It will also allow us to merge either just my changes or (much preferred) both sets of changes at the same time once we get them through review. I don't want his stuff delayed but this means we are not stuck there are any blockers for his code (which is still a possibility given the difficulty of knowing what might turn up in VS). How does that sound?

Reply:
Sounds ok to me. I'd like to get my eyes on those changes as well. Please let me know when there is a PR to look at.

Reply:
Ok, will do.